src/game/logic/object/locomotor.cpp

```diff
@@ -676,60 +676,472 @@ void Locomotor::Loco_Update_Move_Towards_Position(
+
+    if (desired_speed > max_speed) {
+        desired_speed = max_speed;
+    }
+
+    float turn_speed = m_template->m_minTurnSpeed;
+    float orientation = obj->Get_Orientation();
+    float goal_angle = GameMath::Atan2(goal_pos.y - obj->Get_Position()->y, goal_pos.x - obj->Get_Position()->x);
+
+    float turn_angle = Normalize_Angle(goal_angle, orientation);
+    bool backwards = false;
+
+    if (max_speed / 4.0f > turn_speed) {
+        turn_speed = max_speed / 4.0f;
+    }
+
+    float cur_speed = physics->Get_Forward_Speed_2D();
+    bool b = false;
+
+    if (cur_speed == 0.0f) {
+        Set_Flag(MOVING_BACKWARDS, false);
+
+        if (m_template->m_canMoveBackwards && GameMath::Fabs(turn_angle) > DEG_TO_RADF(90.0f)) {
+            Set_Flag(MOVING_BACKWARDS, true);
+            Set_Flag(FLAG_8, obj->Get_Geometry_Info().Get_Major_Radius() * 5.0f < on_path_dist_to_goal);
+        }
+    }
+
+    if (Get_Flag(MOVING_BACKWARDS)) {
+        if (GameMath::Fabs(turn_angle) < DEG_TO_RADF(90.0f)) {
+            backwards = false;
+            Set_Flag(MOVING_BACKWARDS, false);
+        } else {
+            backwards = true;
+            Set_Flag(FLAG_8, obj->Get_Geometry_Info().Get_Major_Radius() * 5.0f < on_path_dist_to_goal);
+            b = Get_Flag(FLAG_8);
+
+            if (!b) {
+                goal_angle = Normalize_Angle(goal_angle, DEG_TO_RADF(180.0f));
+                turn_angle = Normalize_Angle(goal_angle, orientation);
+            }
+        }
+    }
+
+    if (GameMath::Fabs(turn_angle) > DEG_TO_RADF(9.0f) && desired_speed > turn_speed) {
+        desired_speed = turn_speed;
+    }
+
+    if (backwards) {
+        cur_speed = -cur_speed;
+    }
+
+    float speed1 = cur_speed / Get_Braking() + 1.0f;
+    float speed2 = cur_speed / 1.5f * speed1 + cur_speed;
+    float speed3 = speed2;
+
+    if (speed2 < 10.0f) {
+        speed3 = 10.0f;
+    }
+
+    float speed4;
+    float orientation2;
+    float speed5;
+    float speed6;
+    float c;
+    float s;
+    Coord3D pos;
+    Coord3D pos2;
+
+    if (GameMath::Fabs(turn_angle) <= DEG_TO_RADF(15.0f)) {
+        goto l1;
+    }
+
+    speed4 = (desired_speed + cur_speed) * 15.0f / 2.0f;
+    orientation2 = obj->Get_Orientation();
+    speed5 = (desired_speed + cur_speed) / 2.0f / turn_speed;
+
+    if (speed5 > 1.0f) {
+        speed5 = 1.0f;
+    }
+
+    speed6 = 15.0f * speed5 * max_turn / 4.0f;
+    orientation2 = turn_angle >= 0.0f ? orientation2 + speed6 : orientation2 - speed6;
+    c = GameMath::Cos(orientation2) * speed4;
+    s = GameMath::Sin(orientation2) * speed4;
+
+    pos.x = c + obj->Get_Position()->x;
+    pos.y = s + obj->Get_Position()->y;
+    pos.z = obj->Get_Position()->z;
+
+    pos2.x = c / 2.0f + obj->Get_Position()->x;
+    pos2.y = s / 2.0f + obj->Get_Position()->y;
+    pos2.z = obj->Get_Position()->z;
+
+    if (!g_theAI->Get_Pathfinder()->Valid_Movement_Terrain(obj->Get_Layer(), this, &pos2)
```

Comment:
this if isn't right as
```
if ( !Pathfinder::validMovementTerrain(v15, v14, this, &pos2) )
    {
        goto l1;
    }
``

Reply:
Ok, I checked pseudo again and changed it.